cql3/restrictions/statement_restrictions.cc

```diff
@@ -1021,5 +1022,42 @@ ::shared_ptr<single_column_restriction> single_column_restriction::LIKE::apply_t
     return r;
 }
 
+namespace wip {
+
+::shared_ptr<expression> make_conjunction(::shared_ptr<expression> a, ::shared_ptr<expression> b) {
```

Comment:
Let's make expression a shared_ptr<variant<...>> to reduce boilerplate, and add make_expression<type>(args) that we can later point at make_unique instead od make_shared.

Reply:
Or even drop all indirection. If expression is a variant<>, then it is copyable and movable if all its alternatives are copyable/movable resp. So we can just work with value types.

Reply:
> Or even drop all indirection. If expression is a variant<>, then it is copyable and movable if all its alternatives are copyable/movable resp. So we can just work with value types.

I'd love to do that, but I'm stymied by operator_type, which is neither copyable nor assignable. :(  I could build my own version of it, of course.

Reply:
My opinion - go for it. I also got stung by operator_type being a weird kind of singleton, while it could (and should) be a simple enum.

Reply:
Then again, I could always keep a const pointer to operator_type.  That class is written to support such use, whether deliberately or not.

Reply:
Dropped indirection and now constructing expression by value everywhere.  I used a const pointer for operator_type.

Reply:
const* is indeed better IMO. operator_type deserves a rewrite but not now.